### ISSUE TITLE
Support one-to-one mapping of upstream/downstream connections

### ISSUE THREAD

**Issue by chradcliffe** [envoyproxy/envoy#12370]
I’d like to have the ability to have exactly one upstream connection for every downstream connection that we decide to route. That is, streams on the same downstream connection should share the same upstream connection, but no other streams should use that upstream connection.

The rationale for this is twofold:
1. I want to ensure that only the traffic for a single downstream client is carried over a given upstream connection. I do not want mixing of client traffic over an upstream connection. This would be solved by creating a new connection for every stream if it weren't for the second rationale
2. I want to avoid the overhead of the TCP handshake that would be needed in the connection-per-stream case.

I have time to work on this feature if there is a consensus that it makes sense.

**Comment by alyssawilk** [comment]
@mattklein123 @snowp 

I don't think this is going to be an easy one given the design of how we do upstream selection (where it's easy to select the host by all sorts of rules, but there's no transparency on how to select the connection) but I can easily imagine other folks wanting it so I think it's reasonable to land upstream.  WDYT?

**Comment by snowp** [comment]
Yeah I think it will require a bit of work due to the existing abstractions. One way I can imagine doing this is as follows:

- introduce a new conn pool that keys off the downstream connection id, potentially wrapping the existing h1/h2 conn pools (this needs some fleshing out)
- extend the Http::ConnectionPool::newStream function so that additional information can be passed (to pass the connection id)
- Add some connection shutdown cleanup logic, so that when the downstream connection closes the corresponding upstream connection is torn down

**Comment by mattklein123** [comment]
Do we even need a new conn pool for this? Couldn't we use the existing hash functionality in conn pool selection to make sure that a different pool is created for each downstream connection?

**Comment by snowp** [comment]
I think the hash functionality is just for selecting which host to route to, the connection pools will always just attach the request to the first ready connection for the selected host.

**Comment by mattklein123** [comment]
The hash functions go beyond that, they allow separating pools by many things including downstream protocol, etc. If we separate them by downstream client, I think we guarantee what is requested?

**Comment by snowp** [comment]
Ah sorry yeah I conflated this with the LB hashing. I think that would work, though some work might be required to ensure that unused conn pools get cleaned up? Otherwise this would just grow until the max_conn_pools cb is hit or the process ooms.

**Comment by mattklein123** [comment]
>Ah sorry yeah I conflated this with the LB hashing. I think that would work, though some work might be required to ensure that unused conn pools get cleaned up? Otherwise this would just grow until the max_conn_pools cb is hit or the process ooms.

Yeah, this would be an issue, though it might be nice to solve this problem anyway as I think every use of this method has the same issue.

**Comment by chradcliffe** [comment]
Would there be a significant memory/data structure size cost to having a pool per downstream connection? The original suggestion (as I understand it) of keeping a id-to-upstream-connection map in a pool seems like it would be more memory efficient. I don't have the firmest grasp on the memory overhead introduced per pool, so it could be a non-issue.

**Comment by mattklein123** [comment]
I haven't measured it but I don't expect the overhead of a pool to be that high. It would certainly be a lot simpler than a new conn pool impl (modulo the "leak" issue that Snow raises), but feel free to take a look at both options.

**Comment by chradcliffe** [comment]
I'm starting to look at how we could ensure that connection pools are cleaned up. There are two general approaches I can think of:

1. For my current use case, I think the best option would be to clean up the pool when the downstream connection closes. This obviously doesn't solve the general case; however.

2. Another option that would solve the general case would be to have some sort of idle timeout on the entries. The idle timeout has the downside of potentially leaving pools open longer than needed. It also has the downside of tearing down too soon in the case where the downstream connection is not processing stream data but is being kept alive -- this would cause multiple upstream connections to be created for each downstream connection that exhibits this idling behaviour.

(1) could be achieved by registering for `ConnectionCallbacks` with the downstream connection and listening for the close events. This registration seems to make the most sense inside `ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::connPool`, but I think in that context we don't have access to a non-const `Connection*`. We can access a downstream `const Connection*` through the `LoadBalancerContext`, but that's not enough to register for callbacks (and the `LoadBalancerContext` afaik gets its `const Connection*` from the decoder callbacks, which don't expose a non-const downstream `Connection*`. Is there a better way to do cleanup on connection close in this context?

I've thought the most about (1) because it seems better for my current use case, but if (2) is the more acceptable solution I can do some thinking around that. Of course, there may be another option that I haven't thought of at all.

**Comment by mattklein123** [comment]
I think (2) is a more general solution that would likely work for all of the other reasons we might make a connection pool via hashing. I think (2) could potentially become safer also if we only GC it if the pool itself doesn't have any connections. This would encourage configuring idle timeouts on pool connections. WDYT?

**Comment by chradcliffe** [comment]
That seems reasonable, but I'd just like to check my understanding before proceeding: we'd like to refcount `Http::ConnectionPool::Instance` and schedule an idle timer to delete the pool unless another reference is acquired. This approach will ensure that the pools will be GC'd after being idle with no associated downstream stream (or other user of the pool). This, however, won't guarantee that a downstream connection will get the same pool after being idle longer than the upstream pool's idle timeout. Do I have that right?

Question: is there a generic refcounting pointer in the codebase that  (e.g. like `boost::intrusive_ptr`)? I see `Stats::RefcountPtr` but since that's in the `Stats` namespace I'm not sure if that's intended to be general-purpose.

**Comment by mattklein123** [comment]
> we'd like to refcount Http::ConnectionPool::Instance and schedule an idle timer to delete the pool unless another reference is acquired. This approach will ensure that the pools will be GC'd after being idle with no associated downstream stream (or other user of the pool).

What about just have an `onIdle()` callback? Then the cluster manager could schedule the timer and check if it's actually idle before deleting? I think this might be simpler than dealing with ref counting?

> This, however, won't guarantee that a downstream connection will get the same pool after being idle longer than the upstream pool's idle timeout. Do I have that right?

Right, if the downstream idle is larger then the upstream idle it could lead to a new pool/connection in this implementation.

**Comment by chradcliffe** [comment]
Can we get this reopened? #13061 should resolve this, but it's a WIP.

**Comment by denniskniep** [comment]
Hi,

is this issue already fully solved?

I see these PRs are merged:
1. https://github.com/envoyproxy/envoy/pull/12812
> If connection_pool_per_downstream_connection is true, the cluster will use a separate connection pool for every downstream connection

2. https://github.com/envoyproxy/envoy/pull/16948 
https://github.com/envoyproxy/envoy/pull/20998
> Delete connection pools when they have no connections anymore. This fixes unbounded memory use for cases where a new connection pool is needed for each downstream connection

Is there something missing to close this issue?
